database/src/main/java/com/firebase/ui/database/FirebaseArrayOfObjects.java

```diff
@@ -0,0 +1,231 @@
+        return mParser.parseSnapshot(mSnapshots.get(index));
+    }
+
+    @Override
+    public int indexOf(Object o) {
+        return getObjects().indexOf(o);
+    }
+
+    @Override
+    public int lastIndexOf(Object o) {
+        return getObjects().lastIndexOf(o);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return an immutable list iterator
+     */
+    @Override
+    public ListIterator<E> listIterator() {
+        return new ImmutableListIterator(getObjects().listIterator());
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return an immutable list iterator
+     */
+    @Override
+    public ListIterator<E> listIterator(int index) {
+        return new ImmutableListIterator(getObjects().listIterator(index));
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
+
+        FirebaseArrayOfObjects<?> array = (FirebaseArrayOfObjects<?>) obj;
+
+        return mSnapshots.equals(array.mSnapshots) && mEClass.equals(array.mEClass);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = mSnapshots.hashCode();
+        result = 31 * result + mEClass.hashCode();
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return "FirebaseArrayOfObjects{" +
+                "mSnapshots=" + mSnapshots +
+                '}';
+    }
+
+    protected static class FirebaseArrayOfObjectsOptimized<E> extends FirebaseArrayOfObjects<E>
+            implements ChangeEventListener, SubscriptionEventListener {
+        protected List<E> mObjects = new ArrayList<>();
+
+        public FirebaseArrayOfObjectsOptimized(FirebaseArray snapshots, Class<E> modelClass) {
+            super(snapshots, modelClass);
+            snapshots.addChangeEventListener(this);
+            snapshots.addSubscriptionEventListener(this);
+        }
+
+        @Override
+        protected List<E> getObjects() {
+            return mObjects;
+        }
+
+        @Override
+        public void onChildChanged(ChangeEventListener.EventType type, int index, int oldIndex) {
+            switch (type) {
+                case ADDED:
+                    mObjects.add(get(index));
+                    break;
+                case CHANGED:
+                    mObjects.set(index, get(index));
+                    break;
+                case REMOVED:
+                    mObjects.remove(index);
+                    break;
+                case MOVED:
+                    mObjects.add(index, mObjects.remove(oldIndex));
+                    break;
+            }
+        }
+
+        @Override
+        public void onSubscriptionRemoved() {
+            FirebaseArray snapshots = (FirebaseArray) mSnapshots;
+            if (!snapshots.isListening()) {
+                snapshots.removeChangeEventListener(this);
+            }
+        }
+
+        @Override
+        public void onSubscriptionAdded() {
```

Comment:
I'm having a hard time wrapping my head around the SubscriptionEventListener concept.  It seems to be a meta-listener, letting you know when a ChangeEventListener has been added or removed from a source.

Here the logic is "if a listener is added to my FirebaseArray and I am not listening to changes on that array, add me as a listener" ... it seems very self-referential and I am trying to figure out the problem it solves.